README.md
Initialisation
==============

L'initialisation du framework passe par l'appel de la fonction renvoyée par require
avec le dossier racine du projet en argument (celui dans lequel se trouve le
dossier `config`)

```javascript
require('lassi')(__dirname+'/..);
```

L'étape suivante consiste en la création du composant principal

```javascript
var monApplication = lassi.component('MonAppli');
```

On peu ensuite rattacher des services au composant :

```javascript
monApplication.service('NomService', function() { ... })
```

Le service sont globaux. Une fois enregistrés, ils sont accessibles de tous les
composants. Attention donc aux conflits de nommage.

Une fois un service enregistré, il peut être injecté par exemple dans un autre service en
indiquant son nom en paramètre de la fonction de callback

```javascript
monComposant.service('NomAutreService', function(NomService) { ... })
```

On peut aussi rajouter une entité à un composant. Une entité est un cas particulier de
service :

```javascript
monComposant.entity('NomEntité', function() {
  this.construct(function() {
    this.champ = 'default value';
  })
  this.beforeLoad(function(cb) {
    this.champ = ;
    cb();
  })
  this.afterLoad(function(cb) {
    this.champ = ;
    cb();
  })
  this.afterStore(function(cb) {
    this.champ = ;
    cb();
  })
  this.defineIndex('champ', 'integer', function() { });
})
```

Elle peut donc aussi être injectée :

```javascript
monComposant.service('NomAutreService', function(NomService, NomEntité) { ... })
```

Pour paramétrer un composant, on peut lui adjoindre un configurateur

```javascript
monComposant.config(function() { /* code exécuté à l'init du composant */ })
```

On peut ensuite rajouter des controllers qui peuvent avoir des dépendances

```javascript
monApplication.controller('chemin/du/controleur', function(MonService, MonEntité) {

  // une première action pour la route chemin/du/controleur/cheminAction/*/* en get
  this.get('cheminAction/:foo/:bar', function(context) {
    // le code de cette action, on peut récupérer les paramètres de la route passée en 1er argument
    var myFoo = context.arguments.foo;
    var myBar = context.arguments.bar;
    // pour utiliser un transport (en fait rendu) défini ailleurs
    context.next(null, {hello: 'world', foo:myFoo, bar:myBar});
    // pour imposer ici un rendu html (à la place de la ligne précédente)
    context.html({
      // le dossier dans lequel on cherchera les vues
      $views: __dirname+'/views',
      // les infos à mettre dans le <head>
      $metas : {
        title : 'Mon titre de page',
        css   : ['styles/main.css'],
        js    : ['vendors/jquery.min.js'],
      },
      // le layout à utiliser (dans le dossier views)
      $layout : 'layout-page',
      // les sections, chaque propriété qui ne commence pas par $ sera envoyé à la vue)
      // le rendu de la section sera envoyé au layout (comme valeur d'une variable ayant le nom de la section)
      section1 : {
        // le nom de la vue dans $views, si absent on prendra le nom de la section
        $view : 'foo',
        // les propriétés envoyées à la vue
        bar : 'une valeur'
      },
      baz : 'autre valeur' // variable baz passée directement au layout sans rendu si ce n'est pas un objet
    })
  });
})
```

`controller()` peut prendre en premier paramètre un chemin racine d'url, auquel cas ce chemin
sera utilisé pour les actions qu'il contient. De même l'action peut avoir un chemin en
premier paramètre et il sera ajouté au chemin du contrôleur.

Dans le contrôleur, les fonctions appelées par this.XXX correspondent aux méthodes HTTP
(get,put,delete,post,options,etc.). Un cas spécial est `serve` qui permet de publier
un dossier complet.

```javascript
// va rendre accessible le fichier public/foo.bar via l'url /une/racine/un/sous/dossier/foo.bar
monApplication.controller('une/racine', function() {
  this.serve('un/sous/dossier', __dirname+'/public');
})
```

Et pour démarrer l'application on appelle bootstrap() sur le composant principal

```javascript
monApplication.bootstrap();
```

## tests

```
mocha --full-trace -b test/entities.js
```


CHANGELOG.md
Version 2
=========
Version 2.2 (redis)
-------------------
* 2.2.11 (2018-01-17)
  * [BREAKING CHANGE] commande purge renommé en purgeDeleted (et optimisée)
  * dépendances obsolètes virées (mime, mkdirp, mysql, static-favicon, validator)
* 2.2.10 (2018-01-04)
  * solidification countBy en imposant un index existant
  * méthodes internes sorties de la classe EntityQuery (alterLastMatch, createEntitiesFromRows, getType, hasIndex, prepareRecord)
  * les index non lassi d'une collection lassi sont tous virés à l'init de l'Entity
  * Ajout de Entities.close()
  * Entities.dropIndexes() supprimé (deprecated depuis longtepms)
  * Entities.initializeEntity() supprimé (doublon de EntityDefinition.initialize)
  * refacto des tests pour les rendre plus indépendants
  * ajout de $cache.quit() et refacto du test pour l'utiliser (et que mocha rende la main à la fin de ce test seul)
  * passage à eslint standard et fix de tous les warnings qui traînaient (certains depuis longtemps)
* 2.2.9 (2017-12-29)
  * ajout countBy
* 2.2.8 (2017-12-29)
  * ajout updates non bloquants
* 2.2.7 (2017-12-07)
  * ajout maxAge: 1h par dégitfaut (pour serve uniquement)
* 2.2.6 (2017-11-24)
  * ajout d'un accessLog si précisé en config
* 2.2.5 (2017-11-24)
  * Modif body-parser (on prend celui d'express, ajout d'un wrapper pour parser json ou urlencoded suivant contexte mais pas les deux)
  * ajout d'un railErrorHandler pour gérer une erreur d'un middleware
* 2.2.4 (2017-11-20)
  * modifications mineures, nettoyage, commentaires…
* 2.2.3 (2017-11-06)
  * fix $cache.del, code avec cb et promisify si pas de cb
* 2.2.2 (2017-11-03)
  * modif des logs cli, modifs mineures
* 2.2.1 (2017-11-03)
  * amélioration des tests redis
* 2.2.0 (2017-11-02)
  * le cache utilise redis (et plus memcached)

Version 2.1 (mongoDb)
---------------------
* 2.1.24 (2017-10-30)
  * Ajout des auto-updates
  * export de HARD_LIMIT_GRAB
* 2.1.23 (2017-10-16)
  * refacto du store, avec divers fixes : cast est maintenant le même dans Entity et EntityDefinition, fix affectation conditionnelle de __deletedAt, fix removeTemporaryFields
* 2.1.22 (2017-10-12)
  * fix cast sur valeurs absentes
* 2.1.21 (2017-10-10)
  * lassi plus souple sur query.in([]) (log une erreur mais plante plus)
* 2.1.20 (2017-10-10)
  * ajout de $byPassDuplicate et élimination des $properties
* 2.1.19 (2017-10-05)
  * ajout Entity.onLoad()
* 2.1.18 (2017-10-04)
  * upgrade express & body-parser
* 2.1.17 (2017-09-20)
  * On durcit avec un plantage au in|notIn ou avec greaterThan & co si la valeur passée est invalide (undefined ou tableau vide pour in par ex)
* 2.1.16 (2017-09-16)
  * blindage contre des datas invalides
  * aj de getCookie et setCookie
  * ajout d'un context.setNoCache, que l'on utilise sur les redirect (sauf les 301)
* 2.1.15 (2017-09-15)
  * suppression du warning si limit = hardLimit
  * fix sur la création d'index
  * ajout de l'initialize des entities (qui était tout simplement ignoré)
* 2.1.14 (2017-09-13)
  * ajout de Context.restKo()
* 2.1.13 (2017-09-12)
  * ajout de EntityQuery.purge()
  * aj de markToRestore, pour masquer la propriété __deletedAt aux applis
  * tests mongo réparés (ok aussi sur mongo 3.4)
* 2.1.12 (2017-09-08)
  * fix softDelete in Entity#store()
  * ajout EntityQuery#includeDeleted() & Entity#isDeleted()
  * ajout EntityQuery#notEquals()
  * fix flush si la collection n'existe pas
* 2.1.11 (2017-08-31)
  * ajout d'un log.error si on remonte hardLimit résultats
* 2.1.10 (2017-08-31)
  * hardLimit monté à 5000, fix sur l'avertissement qui pouvait donner du faux positif
* 2.1.9 (2017-08-29)
  * ok remplacé par success sur les appels rest
* 2.1.8 (2017-08-25)
  * full text search
  * aj d'une purge générique en cli
* 2.1.7 (2017-08-18)
  * on autorise match _id
  * amélioration des messages d'erreur
* 2.1.6 (2017-08-02)
  * merge des modifs de la branche mysql (2.0) qui a évolué en // (2.0.14 à 2.0.18)
* 2.1.5 (2017-08-02)
  * Ajout soft-delete
  * ajout d'un hardLimit à grab, et on accepte skip ou offset en options
  * Ajout du middleware maintenance
  * pass obligatoire si user mongo (sinon ni l'un ni l'autre)
* 2.1.4 (2017-04-28)
  * Les timeout par défaut deviennent des constantes de lassi
* 2.1.3 (2017-04-13)
  * Add beforeDelete() trigger
  * fix isNull() et index sur date indéfinie
* 2.1.2 (2017-04-11)
  * Fix EntityQuery#notIn + allow multiple match on same attribute
* 2.1.1 (2017-03-29)
  * Fix dependencies
* 2.1.0 (2017-03-29)
  * Première version d'une branche mongo

Version 2.0 (dernière version avec mysql)
-----------------------------------------
* 2.0.18 (2017-06-20)
  * modif des logs cli
  * amélioration du message d'erreur quand la vue n'existe pas
* 2.0.17 (2017-05-16)
  * modifs de log, améliorations cli
* 2.0.16 (2017-02-07)
  * ajout de $entities-cli
* 2.0.15 (2017-01-13)
  * ajout de tests unitaires sur entities
* 2.0.14 (2017-01-13)
  * aj debug sur count
* 2.0.13 (2016-12-16)
  * aj d'options à grab pour permettre un distinct
* 2.0.12 (2016-12-15)
  * blindage avec cast sur chaque index
* 2.0.11 (2016-12-01)
  * fix mineurs (variables inutilisées)
…
* 2.0.0 (2016-08-26)

Version 1
=========
* 1.8.64 (2016-04-12)
…
* 1.8.1
  - Suppression des CORS dans lassi, à gérer par l'appli
  - Ajout d'une méthode "use" à l'application pour injecter ses propres middleware (voir l'usage de railUse pour plus d'info).
* 1.8.0
  - Controller : Les fonctions get/put/etc retourne un le controller et non l'action
  - Controller : JSON est le format par défaut
  - Action : Controller('XXX').get(cb) => XXX comme chemin pour l'action
  - Controller : l'absence de chemin (undefined) => / (racine)
  - Application : Ajout du parsing BODY/JSON et d'un reviver d'objet Date

* 1.7.25 (2015-01-21)
  * init du repository github


source/Lassi.js
'use strict'

/*
 * @preserve
 * This file is part of "lassi".
 *    Copyright 2009-2015, arNuméral
 *    Author : Yoran Brault
 *    eMail  : yoran.brault@arnumeral.fr
 *    Site   : http://arnumeral.fr
 *
 * "lassi" is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * "lassi" is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with "lassi"; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the fs. site: http://www.fs..org.
 */

const flow = require('an-flow')
const _ = require('lodash')
const Component = require('./Component')
const Services = require('./tools/Services')
const EventEmitter = require('events').EventEmitter
const fs = require('fs')
const log = require('an-log')('lassi')
// ajoute les propriétés de couleur sur les string ('toto'.blue pour l'afficher bleu si y'a un tty)
require('colors')
const requireContext = this

let shutdownRequested = false

/**
 * Constructeur de l'application. Effectue les initialisations par défaut.
 * Ce constructeur n'est jamais appelé directement. Utilisez {@link lassi.Lassi}
 * @constructor
 * @param {String=} root La racine du projet. Par défaut il s'agit du dossier d'exécution du script.
 * @extends Emitter
 */
class Lassi extends EventEmitter {
  constructor (options) {
    super()
    // pas mal de taf pour pouvoir s'en passer…
    // if (!options.noGlobalLassi)
    global.lassi = this

    this.transports = {}
    const HtmlTransport = require('./transport/html')
    const JsonTransport = require('./transport/json')
    const RawTransport = require('./transport/raw')
    /**
     * Liste de transports (html, json et raw au bootstrap,
     * avec les alias 'text/html', 'application/json' et 'text/plain')
     */
    this.transports.html = new HtmlTransport(this)
    this.transports.json = new JsonTransport(this)
    this.transports.raw = new RawTransport(this)
    this.transports['text/plain'] = this.transports.raw
    this.transports['text/html'] = this.transports.html
    this.transports['application/json'] = this.transports.json
    this.transports['application/javascript'] = this.transports.raw

    this.components = {}
    this.services = new Services()
    lassi.options = options
    // cette commande nous ajoute lassi en component
    const lassiComponent = lassi.component('lassi')
    lassiComponent.service('$settings', require('./services/settings'))
    lassiComponent.service('$cache', require('./services/cache'))
    lassiComponent.service('$entities', require('./services/entities'))
    lassiComponent.entity('LassiUpdate', require('./updates/LassiUpdate'))
    lassiComponent.service('$updates', require('./services/updates'))
    lassiComponent.service('$maintenance', require('./services/maintenance'))
    if (lassi.options.cli) {
      lassiComponent.service('$cli', require('./services/cli'))
      lassiComponent.service('$cache-cli', require('./services/cache-cli'))
      lassiComponent.service('$entities-cli', require('./services/entities-cli'))
      lassiComponent.service('$maintenance-cli', require('./services/maintenance-cli'))
      lassiComponent.service('$updates-cli', require('./services/updates-cli'))
    } else {
      lassiComponent.service('$rail', require('./services/rail'))
      lassiComponent.service('$server', require('./services/server'))
    }

    options.root = fs.realpathSync(options.root)
    lassi.settings = options.settings ? options.settings : require(options.root + '/config')
    lassi.settings.root = options.root
    // On ajoute un basePath s'il n'existe pas (le préfixe des routes pour des uri absolues)
    if (!lassi.settings.basePath) lassi.settings.basePath = '/'
    // et les composants par défaut qui seront mis en premier (seulement lassi lui-même
    // mis par le lassi.component ci-dessus, mais on laisse ça au cas où qqun ajouterait
    // des components dans ce constructeur)
    this.defaultDependencies = _.keys(lassi.components)
  }

  startup (component, cb) {
    if (!cb) cb = (error) => error && console.error(error)
    log('startup component', component.name)
    const lassiInstance = this
    component.dependencies = this.defaultDependencies.concat(component.dependencies)
    flow()
    // Configuration des composants
      .seq(function () {
        component.configure()
        this()
      })
    // Configuration des services
      .seq(function () {
      /**
       * résoud un service à l'ajoute à setupables et postSetupable si besoin
       * @private
       * @param {string} name
       */
        function addService (name) {
          if (added.has(name)) return
          added.add(name)
          const service = lassiInstance.services.resolve(name) // Permet de concrétiser les services non encore injectés
          if (!service) throw new Error(`Le service ${name} n'a pu être résolu (il ne retourne probablement rien)`)
          if (service.setup) setupables.push(service)
          if (service.postSetup) postSetupable.push(service)
        }
        // pour mémoriser les services déjà ajoutés
        /* global Set */
        const added = new Set()
        // liste des setup à lancer
        const setupables = []
        // postSetup est utilisé pour les services qui ont des actions à faire quand tous les services
        // sont dispo (setup terminé, par ex pour que la BDD soit initialisée)
        // mais avant $server.start()
        const postSetupable = []
        const services = lassiInstance.services.services()
      // on veut $settings puis $cache puis $entities dispos dans cet ordre,
      // pour que les autres setup puissent les utiliser
      ;['$settings', '$cache', '$entities'].forEach(addService)
        Object.keys(services).forEach(addService)
        // fin init
        log('starting setup chain', Array.from(added).join(', '))
        flow(setupables)
          .seqEach(function (service) {
            service.setup(this)
          })
          .seq(function () {
            log('starting postSetup chain')
            this(null, postSetupable)
          })
          .seqEach(function (service) {
            log('postSetup', service.serviceName)
            service.postSetup(this)
          })
          .done(this)
      })
      .seq(function () {
        log('startup event')
        lassiInstance.emit('startup')
        cb()
      })
      .catch(cb)
  }

  /**
   * Démarre l'application d'un composant.
   * @param {Component} component Le composant
   * @private
   */
  bootstrap (component, cb) {
    const lassiInstance = this
    flow()
      .seq(function () {
        lassiInstance.startup(component, this)
      })
      .seq(function () {
        if (lassiInstance.options.cli) return this()
        const $server = lassiInstance.service('$server')
        $server.start(this)
      })
      .done(function (error) {
        if (error) console.error(error.stack)
        if (cb) cb(error)
      })
  }

  /**
   * Enregistre un {@link Component} dans le système.
   * @param {String} name           Le nom du component
   * @param {string[]}  [dependencies] Une liste de noms de composants en dépendance
   */
  component (name, dependencies) {
    this.components[name] = new Component(name, dependencies)
    return this.components[name]
  }

  /**
   * Enregistre un {@link Service} dans le système.
   * @param {String} name           Le nom du component
   */
  service (name) {
    return this.services.resolve(name)
  }

  /**
   * Liste tous les services enregistrés
   */
  allServices () {
    return this.services.services()
  }

  /**
   * Arrêt de l'application.
   * @private
   * @fires Lassi#shutdown
   */
  shutdown (exitCode = 0) {
    function thisIsTheEnd () {
      log('shutdown completed (exit in 1s)')
      // on ajoute quand même 1s pour passer après des process.nextTick lancés sur l'event shutdown
      // et laisser les connexions se faire ou se terminer
      // (envoi de mail de notif par ex)
      setTimeout(() => process.exit(exitCode), 1000)
    }

    // normal d'être appelé 2× avec SIGINT puis exit
    if (shutdownRequested) return

    shutdownRequested = true
    log('processing shutdown')
    // Avant de lancer l'événement on met une limite pour les réponses à 2s
    setTimeout(function () {
      log('shutdown too slow, forced')
      thisIsTheEnd()
    }, 2000)

    /**
     * Évènement généré lorsque l'application est arrêtée par la méthode shutdown.
     * @event Lassi#shutdown
     */
    this.emit('shutdown')

    process.nextTick(() => {
      try {
        // Dans certains cas, this.service n'existe déjà plus !
        if (this.service) {
          // Avec un sgbd, il fallait pas fermer la connexion ici
          // sinon les transactions en cours ne pouvaient pas se terminer
          log('closing $entities')
          this.service('$entities').close()

          // on ferme aussi $cache (ça ferme le client redis de la session qui est le même)
          log('closing redis connection')
          this.service('$cache').quit()

          // et pour finir $server, si on est pas en cli…
          if (lassi.options.cli) {
            thisIsTheEnd()
          } else {
            log('closing $server')
            this.service('$server').stop(thisIsTheEnd)
          }
        } else {
          log('server is already gone')
          thisIsTheEnd()
        }
      } catch (error) {
        log.error('error on shutdown', error)
        process.exit()
      }
    })
  }
}

/**
 * Démarre lassi avec les options fournies
 * @param {string|object} options Si string interprété comme options.root
 * @param {boolean} [options.cli=false] Passer true pour activer les services *-cli (et ne pas lancer le serveur http)
 * @param {boolean} [options.noGlobalLassi=false] @todo Passer true pour que lassi ne soit pas mis en global
 * @param {string} options.root la racine de l'appli
 * @param {object} [options.settings] Passer un objet de settings complet (sinon ira chercher ${root}/config.js)
 * @return {Lassi}
 */
function startLassi (options) {
  // on accepte root tout seul à la place de l'objet options
  if (typeof options === 'string') {
    options = {
      root: options
    }
  }
  options.cli = !!options.cli
  // on ne tolère qu'un seul lassi en global, mais on pourrait en avoir plusieurs
  // en passant du options.noGlobalLassi = true (quand lassi lui-même pourra se passer d'être global)
  if (_.has(global, 'lassi')) {
    log('ERROR'.red, ' : lassi already exists')
    return global.lassi
  }
  const lassi = new Lassi(options)
  lassi.Context = require('./Context')
  lassi.require = () => require.apply(requireContext, arguments)

  // Un écouteur pour tout ce qui pourrait passer au travers de la raquette
  // @see https://nodejs.org/api/process.html#process_event_uncaughtexception
  process.on('uncaughtException', (error) => {
    // On envoie l'erreur en console
    console.error('uncaughtException : ', error)
  })

  // On ajoute nos écouteurs pour le shutdown car visiblement beforeExit n'arrive jamais, et exit ne sert
  // que sur les sorties "internes" via un process.exit() car sinon on reçoit normalement un SIG* avant
  if (!options.cli) {
    // see https://en.wikipedia.org/wiki/Signal_(IPC)#POSIX_signals
    // ctrl + c => SIGINT
    // fermeture du term parent => SIGHUP
    // kill -N pid, avec N :
    // --------------- 1 ------ 2 ------- 15
    ['beforeExit', 'SIGHUP', 'SIGINT', 'SIGTERM', 'exit'].forEach((signal) => {
      process.on(signal, () => {
        log('pid ' + process.pid + ' received signal ' + signal)
        lassi.shutdown()
      })
    })
  }

  // Le message 'shutdown' est envoyé par pm2 sur les gracefulReload
  process.on('message', (message) => {
    // On récupère bien la string 'shutdown' qui est affichée ici
    log(`message "${message}" of pid ${process.pid}`)
    if (message === 'shutdown') {
      // Mais on n'arrive jamais là, le process meurt visiblement avant
      log('launching shutdown')
      lassi.shutdown()
    }
  })

  return lassi
}

module.exports = startLassi
